tls/s2n_handshake_io.c

```diff
@@ -688,7 +744,28 @@ int s2n_negotiate(struct s2n_connection *conn, s2n_blocked_status * blocked)
         /* Flush any pending I/O or alert messages */
         GUARD(s2n_flush(conn, blocked));
 
-        if (ACTIVE_STATE(conn).writer == this) {
+        if (ACTIVE_STATE(conn).writer == 'A') {
+            /* We are in a state that is blocked on application data */
+            *blocked = S2N_BLOCKED_ON_APPLICATION_DATA;
+
+            int r = handshake_handle_app_data(conn);
+            if (r < 0) {
+                if (s2n_errno != S2N_ERR_BLOCKED && s2n_allowed_to_cache_connection(conn) && conn->session_id_len) {
+                    conn->config->cache_delete(conn->config->cache_delete_data, conn->session_id, conn->session_id_len);
+                }
+                return -1;
```

Comment:
s2n_negotiate is a public API, when error happened, shall we call S2N_ERROR to let the caller know the reason?

Reply:
S2N_ERROR(x) is to set s2n_errno with x, but at this point, s2n_errno had already been correctly set, see line 753 which we use it to check some conditions.